lib/ansible/modules/files/copy.py

```diff
@@ -483,6 +483,9 @@ def copy_common_dirs(src, dest, module):
         left_only_changed = copy_left_only(b_src_item_path, b_dest_item_path, module)
         if diff_files_changed or left_only_changed:
             changed = True
+
+        # recurse into subdirectory
+        changed = changed or copy_common_dirs(os.path.join(src, item), os.path.join(dest, item), module)
```

Comment:
I'm concerned about introducing a recursive call here. It may be better to enumerate the directory with `os.walk()` rather than making this recursive.

Reply:
The code above this doesn't just enumerate the directory, it compares two directories. This comparison happens *shallowly*, as mentioned in the [documentation of filecmp.dircmp](https://docs.python.org/3/library/filecmp.html#filecmp.dircmp)

I'm not sure how `os.walk()` would help here without basically rewriting `filecmp.dircmp`, since `os.walk()` returns all files, not just the changed files. What's your suggestion?

Also note that the example in the [documentation of filecmp.dircmp](https://docs.python.org/3/library/filecmp.html#filecmp.dircmp) specifically uses recursion:

```python3
from filecmp import dircmp
def print_diff_files(dcmp):
    for name in dcmp.diff_files:
        print("diff_file %s found in %s and %s" % (name, dcmp.left,
              dcmp.right))
    for sub_dcmp in dcmp.subdirs.values():
        print_diff_files(sub_dcmp)

dcmp = dircmp('dir1', 'dir2') 
print_diff_files(dcmp) 
```